cmd/server/server.go

```diff
@@ -0,0 +1,211 @@
+	defaultServerPort               = 8080
+	defaultName                     = "default"
+	defaultInitialAdvertisePeerURLs = "http://localhost:2380"
+)
+
+var logger = logrus.New()
+
+// configuration is struct to hold configuration for utility
+type configuration struct {
+	schedule              string
+	etcdEndpoints         string
+	storageProvider       string
+	maxBackups            int
+	etcdConnectionTimeout int
+}
+
+var (
+	port int
+
+	//restore flags
+	restoreCluster      string
+	restoreClusterToken string
+	restoreDataDir      string
+	restorePeerURLs     []string
+	restoreName         string
+	skipHashCheck       bool
+	storageProvider     string
+)
+
+// NewServerCommand create cobra command for snapshot
+func NewServerCommand(stopCh <-chan struct{}) *cobra.Command {
+	config := &configuration{}
+	var command = &cobra.Command{
+		Use:   "server",
+		Short: "start the http server with backup scheduler.",
+		Long:  `Server will keep listening for http request to deliver its functionality through http endpoins.`,
+		Run: func(cmd *cobra.Command, args []string) {
+
+			clusterUrlsMap, err := types.NewURLsMap(restoreCluster)
+			if err != nil {
+				logger.Fatalf("failed creating url map for restore cluster: %v", err)
+			}
+			peerUrls, err := types.NewURLs(restorePeerURLs)
+			if err != nil {
+				logger.Fatalf("failed parsing peers urls for restore cluster: %v", err)
+			}
+
+			options := &restorer.RestoreOptions{
+				RestoreDataDir: restoreDataDir,
+				Name:           restoreName,
+				ClusterURLs:    clusterUrlsMap,
+				PeerURLs:       peerUrls,
+				ClusterToken:   restoreClusterToken,
+				SkipHashCheck:  skipHashCheck,
+			}
+
+			etcdInitializer := initializer.NewInitializer(options, storageProvider, logger)
+			handler := &server.HTTPHandler{
+				Port:            port,
+				EtcdInitializer: *etcdInitializer,
+				Logger:          logger,
+			}
+			logger.Info("Regsitering the http request handlers...")
+			handler.RegisterHandler()
+			logger.Info("Starting the http server...")
+			go handler.Start()
+
+			defer handler.Stop()
+
+			ss, err := getSnapstore(storageProvider)
+			if err != nil {
+				logger.Fatalf("Failed to create snapstore from configured storage provider: %v", err)
+			}
+			logger.Info("Created snapstore from provider: %s", storageProvider)
+			ssr, err := snapshotter.NewSnapshotter(
+				config.etcdEndpoints,
+				config.schedule,
+				ss,
+				logger,
+				config.maxBackups,
+				time.Duration(config.etcdConnectionTimeout))
+			if err != nil {
+				logger.Fatalf("Failed to create snapshotter: %v", err)
+			}
+			for {
+				logger.Infof("Probing etcd...")
+				select {
+				case <-stopCh:
+					logger.Info("Shutting down...")
+					return
+				default:
+					err = probeEtcd(config.etcdEndpoints, time.Duration(config.etcdConnectionTimeout))
+				}
+				if err == nil {
+					break
+				}
+			}
+			err = ssr.Run(stopCh)
+			if err != nil {
+				logger.Fatalf("Snapshotter failed with error: %v", err)
```

Comment:
We have a cross container `/metrics` liveliness probe. That means if backup fails, etcd should fail. Ideally, backup should fail when etcd fails, right? So what about we get rid of the cross container liveliness probe and keep the liveliness probe of etcd same as the readiness probe. That way, if backup operation fails, sidecar will start probing etcd (which is basically its liveliness probe). When etcd fails the liveliness probe, the container is restarted and sidecar can act on the etcd failure if needed.

Reply:
Even i was little doubtful about this probe, we wanted that behaviour from business perspective, where etcd will also fail if backup fails, right? I agree that cause, container restarts, but we have to consider the worst case scenario as well where sidecar stuck in initial phase itself where it fail to start http server. Although i can't think of any sure shot scenario but just for consideration, rare case could be, just not enough CPU space to start sidecar container but the process. I think is call to make whether to have this probe or not. I'm ready to go without it considering it's rare occasion.